apps/survey/contracts/Survey.sol

```diff
@@ -98,84 +99,104 @@ contract Survey is AragonApp {
-            for (uint256 i = 1; i <= previouslyVoted.optionsCastedLength; i++) {
-                survey.votes[previouslyVoted.castedVotes[i].optionId] = survey.votes[previouslyVoted.castedVotes[i].optionId].sub(previouslyVoted.castedVotes[i].stake);
-            }
-            // remove previous vote participation
-            survey.participation = survey.participation.sub(voterStake.sub(previouslyVoted.castedVotes[0].stake));
-
-            // reset previously voted options, if any
-            delete survey.voters[msg.sender];
-        }
+        // revert previous votes, if any (also checks if canVote)
+        resetVote(_surveyId);
 
         uint256 totalVoted = 0;
-        // reserve first index for NO_VOTE
-        survey.voters[msg.sender].castedVotes[0] = OptionCast({optionId: NO_VOTE, stake: 0});
-        for (uint256 j = 0; j < _options.length; j++) {
-            require(_options[j] != NO_VOTE && _options[j] <= survey.options);
-            require(_stakes[j] > 0);
+        // reserve first index for ABSTAIN_VOTE
+        survey.votes[msg.sender].castedVotes[0] = OptionCast({optionId: ABSTAIN_VOTE, stake: 0});
+        for (uint256 j = 1; j <= _optionIds.length; j++) {
+            // parameter arrays don't have abstain vote, while survey ones do, so we subtract 1 here to the index
+            uint256 optionId = _optionIds[j-1];
+            uint256 stake = _stakes[j-1];
+            require(optionId != ABSTAIN_VOTE && optionId <= survey.options);
+            require(stake > 0);
 
             // register voter amount
-            // index 0 in options array is reserved for NO_VOTE, so index will be 1 more
-            survey.voters[msg.sender].castedVotes[j + 1] = OptionCast({optionId: _options[j], stake: _stakes[j]});
+            // index 0 in options array is reserved for ABSTAIN_VOTE, so index will be 1 more
+            survey.votes[msg.sender].castedVotes[j] = OptionCast({optionId: optionId, stake: stake});
 
             // add to total option support
-            survey.votes[_options[j]] = survey.votes[_options[j]].add(_stakes[j]);
+            survey.optionVotes[optionId] = survey.optionVotes[optionId].add(stake);
 
             // let's avoid repeating an option by
             // making sure that ascending order is preserved in options array
-            // index 0 in options array is reserved for NO_VOTE, so here we are comparing
-            // the option just added (_options[j] = options[j+1]) with the previous one (options[j])
-            require(survey.voters[msg.sender].castedVotes[j].optionId < survey.voters[msg.sender].castedVotes[j + 1].optionId);
-            totalVoted = totalVoted.add(_stakes[j]);
+            // index 0 in options array is reserved for ABSTAIN_VOTE, so here we are comparing
+            // the option just added with the previous one
+            require(survey.votes[msg.sender].castedVotes[j - 1].optionId < survey.votes[msg.sender].castedVotes[j].optionId);
```

Comment:
I think we can actually move this check up to where the other requires are, since we already know `survey.votes[msg.sender].castedVotes[j].optionId`. Then that block will read like we're checking to make sure we can actually add the next option a voter gave.

Reply:
Oops, nope. We don't have `survey.votes[msg.sender].castedVotes[optionIndex]` yet. I just moved it a little bit above.

Reply:
Oops, sorry, I read your comment too fast. You were absolutely right. Thanks for your fix!!!